rfd/0109-opsgenie-plugin.md

```diff
@@ -0,0 +1,101 @@
+
+## Success criteria
+Users are able to configure teleport to automatically create alerts in Opsgenie from access requests.
+Alerts were chosen over incidents in Opsgenie for high priority alerts that indicate a service interruption.
+Users are also able to configure auto approval flows to be met under certain conditions. E.g when a requester is on-call 
+
+## Configuration UX
+
+The plugin will be configured using a toml file containing the required information to interact with both Teleport access and the Opsgenie API.
+
+```
+[opsgenie]
+api_key = "path/to/key" # File containing Opsgenie API Key
+opsgenie_addr = "example.app.opsgenie.com" # Address of Opsgenie
+severity = "2" # Severity to create Opsgenie alerts with
+```
+
+### Getting an Opsgenie API key
+
+In the Opsgenie web UI go to Settings -> App settings -> API key management. Create key with Read, Create and Update access.
+
+### Executing
+
+The plugin will be started using a command of the form
+
+```
+Teleport opsgenie start –config <config file location>
+```
+
+## UX
+
+Once an access request has been created, the Opsgenie plugin will create an alert in the service specified in the request annotation using the Opsgenie Alert API Create endpoint. 
+
+The appropriate on call responder can then click into the provided link and approve or deny the access request.
+
+For auto approval of certain access requests the access request will be auto approved if the requesting user is on-call in one of the services provided in request annotation.
+
+Once an access request has been approved or denied the plugin will add a note to the alert and close the relevant alert tied to that access request.
+
+## Implementation details
+In this section we will take a look at how the plugin will interact with the Opsgenie API.
+
+### Authorization
+
+The plugin will use the API key provided in the Opsgenie config file when interacting with the Opsgenie API. This API key will be included in the headers of the requests made.
+
+```
+Header Key: Authorization
+Header Value: genieKey $apiKey
+```
+
+### Creating alerts
+When the Opsgenie plugin creates alerts for incoming access requests the Create alert request will be of the form
+
+```
+{
+	"message": "Access request from <someuser>",
+	"description":"<someuser> requested permissions for roles <someroles> on Teleport at <someTime>.
+ 	Reason: <someReason>
+ 	To approve or deny the request, proceed to <link to the access request>",
+	"responders":[
+    	....
+	],
+	"tags": ["TeleportAccessRequest"],
+	"priority":"<somePriority>"
+}
+```
+
+When the access request has been approved or denied the alert created in Opsgenie will have a note added to it using the ‘Add note to alert’ endpoint. Then the alert will be closed using the ‘Close alert’ endpoint.
```

Comment:
Mind you, it's possible to configure the access request system so that multiple reviews are required to approve or deny an access request; this should read something along the lines of "on every review, the alert will have a note added to it, and when the access request is approved or denied, the alert will be closed" (it's also possible to deny an access request that was already approved, or to delete an access request at any point).